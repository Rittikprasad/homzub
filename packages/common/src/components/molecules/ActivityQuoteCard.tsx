import React from 'react';
import { FlatList, SectionList, StyleSheet, TouchableOpacity, View } from 'react-native';
import { useTranslation } from 'react-i18next';
import { FunctionUtils } from '@homzhub/common/src/utils/FunctionUtils';
import { StringUtils } from '@homzhub/common/src/utils/StringUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Badge } from '@homzhub/common/src/components/atoms/Badge';
import { Divider } from '@homzhub/common/src/components/atoms/Divider';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import { Avatar } from '@homzhub/common/src/components/molecules/Avatar';
import { Quote } from '@homzhub/common/src/domain/models/Quote';
import { QuoteCategory } from '@homzhub/common/src/domain/models/QuoteCategory';
import { TicketActivityQuote } from '@homzhub/common/src/domain/models/TicketActivityQuote';
import { User } from '@homzhub/common/src/domain/models/User';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

interface IProps {
  quoteData: QuoteCategory[];
  onQuotePress?: (url: string) => Promise<void> | void;
}

interface IQuoteDataItem {
  id: number;
  title: string;
  data: Quote[];
}

interface IQuotesApproved {
  quoteData: TicketActivityQuote[];
  description: string;
}

interface IApprovedQuoteItem {
  id: number;
  title: string;
  price: string;
  user: User;
  role: string;
}

const QuotesSubmitted = (props: IProps): React.ReactElement => {
  const { quoteData, onQuotePress = FunctionUtils.noop } = props;
  const { t } = useTranslation(LocaleConstants.namespacesKey.serviceTickets);

  const formatQuotes = (): IQuoteDataItem[] => {
    return quoteData.map((quoteDataItem) => ({
      id: quoteDataItem.id,
      title: quoteDataItem.name,
      data: quoteDataItem.quotes,
    }));
  };

  const allQuotes = formatQuotes();

  const renderItem = ({ item }: { item: Quote }): React.ReactElement => {
    const {
      id,
      currency: { currencySymbol },
      attachment: { link },
      totalAmount,
      quoteNumber,
    } = item;

    const onPress = async (): Promise<void> => await onQuotePress(link);

    return (
      <TouchableOpacity key={id} style={styles.quoteItem} onPress={onPress}>
        <View style={styles.quoteStart}>
          <Icon name={icons.docFilled} size={20} style={styles.fileIcon} />
          <Label textType="regular" type="large">
            {t('quoteNumber', { number: quoteNumber })}
          </Label>
        </View>
        <Label textType="regular" type="large">{`${currencySymbol}${totalAmount}`}</Label>
      </TouchableOpacity>
    );
  };

  const renderSectionHeader = ({ section }: { section: IQuoteDataItem }): React.ReactElement => {
    const { title } = section;
    const showDivider = allQuotes.indexOf(section) !== 0;
    return (
      <View style={styles.sectionHeader}>
        {showDivider && <Divider containerStyles={styles.divider} />}
        <Label textType="semiBold" type="large">
          {t('quoteCategory', { title })}
        </Label>
      </View>
    );
  };

  const renderItemSeparator = (): React.ReactElement => <View style={styles.itemSeparator} />;

  const keyExtractor = (item: Quote, index: number): string => `${item.id} [${index}]`;

  return (
    <>
      <View style={styles.container}>
        <SectionList
          sections={allQuotes}
          keyExtractor={keyExtractor}
          // @ts-ignore
          renderSectionHeader={renderSectionHeader}
          renderItem={renderItem}
          contentContainerStyle={styles.sectionList}
          ItemSeparatorComponent={renderItemSeparator}
        />
      </View>
      <Badge
        badgeColor={theme.colors.gray11}
        title={t('serviceTickets:awaitingApprovalAndPayment')}
        badgeStyle={styles.activityBadge}
        titleStyle={styles.activityBadgeText}
      />
    </>
  );
};

const QuotesApproved = (props: IQuotesApproved): React.ReactElement => {
  const { quoteData, description } = props;

  const { t } = useTranslation(LocaleConstants.namespacesKey.serviceTickets);
  const formatApprovedQuotes = (): IApprovedQuoteItem[] => {
    return quoteData.map((quote) => {
      const {
        id,
        quoteRequestCategory: { name },
        totalAmount,
        currency: { currencySymbol },
        user,
        role,
      } = quote;
      return {
        id,
        title: name,
        price: `${currencySymbol}${totalAmount}`,
        user,
        role,
      };
    });
  };

  const approvedQuotes = formatApprovedQuotes();

  const keyExtractor = (item: IApprovedQuoteItem): string => `${item.id}:[${item.price}]`;

  const renderItem = ({ item }: { item: IApprovedQuoteItem }): React.ReactElement => {
    const { title, price, role, user } = item;
    const { profilePicture, name } = user;
    return (
      <>
        <Avatar
          image={profilePicture}
          fullName={name}
          designation={StringUtils.toTitleCase(role) || t('property:owner')}
        />
        <View style={styles.approvedQuote}>
          <Label type="large" textType="regular">
            {t('quoteCategory', { title })}
          </Label>
          <Label type="large" textType="semiBold">
            {price}
          </Label>
        </View>
      </>
    );
  };

  const renderItemSeparator = (): React.ReactElement => <Divider containerStyles={styles.approvedDivider} />;

  return (
    <>
      <View style={styles.approvedHeader}>
        <Label textType="semiBold" type="large" style={styles.approvedComment}>
          {t('approvedQuotes')}
        </Label>
        {!!description && (
          <Label textType="regular" type="regular">
            {description}
          </Label>
        )}
      </View>
      <Divider containerStyles={styles.approvedSeparator} />
      <View style={styles.approvedContainer}>
        <FlatList
          data={approvedQuotes}
          renderItem={renderItem}
          keyExtractor={keyExtractor}
          ItemSeparatorComponent={renderItemSeparator}
          contentContainerStyle={styles.approvedFlatlist}
        />
      </View>
      <Badge
        badgeColor={theme.colors.gray11}
        title={t('awaitingWorkInitiation')}
        badgeStyle={styles.activityBadge}
        titleStyle={styles.activityBadgeText}
      />
    </>
  );
};

const ActivityQuotesSubmitted = React.memo(QuotesSubmitted);
const ActivityQuotesApproved = React.memo(QuotesApproved);

export { ActivityQuotesSubmitted, ActivityQuotesApproved };

const styles = StyleSheet.create({
  sectionList: {
    marginHorizontal: 16,
    marginVertical: 20,
  },
  fileIcon: {
    marginEnd: 10,
  },
  divider: {
    marginVertical: 20,
  },
  quoteItem: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  quoteStart: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  sectionHeader: {
    marginBottom: 20,
  },
  container: {
    backgroundColor: theme.colors.gray8,
    marginEnd: 16,
    marginTop: 16,
  },
  itemSeparator: {
    marginTop: 20,
  },
  activityBadge: {
    marginTop: 16,
    minHeight: 26,
    justifyContent: 'center',
    borderRadius: 6,
    marginEnd: 16,
  },
  activityBadgeText: {
    color: theme.colors.darkTint4,
  },
  approvedQuote: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 16,
  },
  approvedDivider: {
    marginVertical: 16,
    borderColor: theme.colors.darkTint10,
  },
  approvedContainer: {
    backgroundColor: theme.colors.gray8,
    paddingVertical: 10,
    marginEnd: 16,
    paddingHorizontal: 16,
  },
  approvedHeader: {
    backgroundColor: theme.colors.gray8,
    marginTop: 20,
    marginEnd: 16,
    paddingHorizontal: 16,
    paddingBottom: 16,
  },
  approvedComment: {
    marginTop: 15,
    marginBottom: 6,
  },
  approvedSeparator: {
    marginEnd: 16,
    borderColor: theme.colors.darkTint10,
  },
  approvedFlatlist: {
    marginVertical: 7,
  },
});
