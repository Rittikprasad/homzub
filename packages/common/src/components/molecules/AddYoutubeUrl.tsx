import React from 'react';
import { View, StyleSheet, ViewStyle, StyleProp, TextInput } from 'react-native';
import { useTranslation } from 'react-i18next';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { useDown } from '@homzhub/common/src/utils/MediaQueryUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import YoutubeSVG from '@homzhub/common/src/assets/images/youtube.svg';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';
import { RNSwitch } from '@homzhub/common/src/components/atoms/Switch';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { deviceBreakpoint } from '@homzhub/common/src/constants/DeviceBreakpoints';

interface IProps {
  isToggled: boolean;
  onToggle: () => void;
  videoUrl: string;
  onUpdateUrl: (url: string) => void;
  containerStyle?: StyleProp<ViewStyle>;
}

export const AddYoutubeUrl = (props: IProps): React.ReactElement => {
  const { isToggled, onToggle, videoUrl, onUpdateUrl, containerStyle } = props;

  const isMobile = useDown(deviceBreakpoint.MOBILE);
  const { t } = useTranslation(LocaleConstants.namespacesKey.property);

  const setPaddingForYouTubeSection = (): number => {
    if (PlatformUtils.isWeb() && !isMobile) return 20;
    if (isMobile) return 10;
    return 0;
  };
  const youtubeSectionPadding = { paddingLeft: setPaddingForYouTubeSection() };
  return (
    <View style={[styles.container, containerStyle]}>
      <View style={styles.header}>
        <Text type="small" textType="semiBold" style={styles.headerText}>
          {t('video')}
        </Text>
        <RNSwitch selected={isToggled} onToggle={onToggle} />
      </View>
      {isToggled && (
        <View style={[styles.youtubeContainer, youtubeSectionPadding]}>
          <View style={styles.youtubeIconAndText}>
            <YoutubeSVG width={35} height={35} />
            <Text type="small" textType="semiBold" style={styles.youtubeUrlText}>
              {t('youtubeUrl')}
            </Text>
          </View>
          <Label type="large" textType="regular" style={styles.helperText}>
            {t('videoHelperText')}
          </Label>
          <TextInput
            textContentType="URL"
            value={videoUrl}
            placeholder={t('urlPlaceholder')}
            keyboardType="default"
            onChangeText={onUpdateUrl}
            style={styles.textInput}
            clearButtonMode="always"
          />
        </View>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    borderRadius: 4,
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderRadius: 4,
    padding: 15,
    backgroundColor: theme.colors.moreSeparator,
  },
  headerText: {
    color: theme.colors.darkTint3,
  },
  youtubeContainer: {
    justifyContent: 'space-around',
    alignItems: PlatformUtils.isWeb() ? 'flex-start' : 'center',
    backgroundColor: theme.colors.white,
    minHeight: 150,
  },
  helperText: {
    color: theme.colors.darkTint5,
  },
  youtubeUrlText: {
    color: theme.colors.darkTint3,
    marginStart: 10,
  },
  textInput: {
    width: 320,
    textAlign: 'left',
    borderWidth: 1,
    borderRadius: 4,
    paddingHorizontal: 14,
    paddingVertical: 12,
    borderColor: theme.colors.disabled,
  },
  youtubeIconAndText: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
});
