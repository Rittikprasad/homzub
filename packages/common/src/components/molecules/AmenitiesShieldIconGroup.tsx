import React, { FC } from 'react';
import { StyleSheet, View } from 'react-native';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import Icon, { icons } from '@homzhub/common/src/assets/icon';

interface IProps {
  iconSize: number;
  onBadgePress: () => void;
  badgesInfo: IBadgeInfo[];
}

interface IBadgeInfo {
  color: string;
}

export const AmenitiesShieldIconGroup: FC<IProps> = ({ onBadgePress, iconSize, badgesInfo }: IProps) => {
  return (
    <View style={styles.badgesContainer}>
      {badgesInfo?.map((item, index) => (
        <Icon
          key={index}
          name={icons.badge}
          size={iconSize}
          color={item.color}
          style={styles.badges}
          onPress={onBadgePress}
        />
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  badgesContainer: {
    flexDirection: 'row',
  },
  badges: {
    marginHorizontal: PlatformUtils.isWeb() ? 0 : 3,
    marginLeft: PlatformUtils.isWeb() ? 4 : 0,
  },
});
