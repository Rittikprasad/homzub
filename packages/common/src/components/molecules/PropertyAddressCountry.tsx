import React from 'react';
import { StyleProp, StyleSheet, TextStyle, View, ViewStyle } from 'react-native';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { ITypographyProps, Typography } from '@homzhub/common/src/components/atoms/Typography';

interface IProps {
  primaryAddress: string;
  isIcon?: boolean;
  subAddress?: string;
  propertyType?: string;
  countryFlag: React.ReactElement | null;
  primaryAddressTextStyles?: ITypographyProps;
  subAddressTextStyles?: ITypographyProps;
  primaryAddressStyle?: StyleProp<TextStyle>;
  subAddressStyle?: StyleProp<TextStyle>;
  containerStyle?: StyleProp<ViewStyle>;
  showAddress?: boolean;
}

const PropertyAddressCountry = (props: IProps): React.ReactElement => {
  const {
    primaryAddress,
    subAddress,
    primaryAddressTextStyles,
    subAddressTextStyles,
    primaryAddressStyle,
    subAddressStyle,
    containerStyle = {},
    countryFlag,
    isIcon,
    showAddress = true,
    propertyType,
  } = props;
  return (
    <View style={[styles.propertyAddressContainer, containerStyle]}>
      {!!propertyType && (
        <Typography variant="label" size="large" fontWeight="regular" style={[styles.propertyType]} numberOfLines={1}>
          {propertyType}
        </Typography>
      )}
      <View style={styles.flexRow}>
        {countryFlag}
        <Typography
          variant={primaryAddressTextStyles?.variant ?? 'text'}
          size={primaryAddressTextStyles?.size ?? 'regular'}
          fontWeight={primaryAddressTextStyles?.fontWeight ?? 'semiBold'}
          style={[styles.propertyNameText, primaryAddressStyle]}
          numberOfLines={1}
        >
          {primaryAddress}
        </Typography>
      </View>
      {showAddress && !!subAddress && (
        <View style={styles.flexRow}>
          {isIcon && <Icon name={icons.locationMarker} size={16} color={theme.colors.darkTint3} style={styles.icon} />}
          <Typography
            variant={subAddressTextStyles?.variant ?? 'text'}
            size={subAddressTextStyles?.size ?? 'small'}
            fontWeight={subAddressTextStyles?.fontWeight ?? 'regular'}
            style={[styles.subAddress, subAddressStyle]}
            numberOfLines={2}
          >
            {subAddress}
          </Typography>
        </View>
      )}
    </View>
  );
};

export { PropertyAddressCountry };

const styles = StyleSheet.create({
  propertyAddressContainer: {
    flexDirection: 'column',
    justifyContent: 'space-between',
  },
  propertyNameText: {
    color: theme.colors.shadow,
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
    marginHorizontal: 12,
  },
  flexRow: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  subAddress: {
    flex: 1,
    color: theme.colors.darkTint3,
    marginVertical: 6,
  },
  flagStyle: {
    borderRadius: 2,
    width: 24,
    height: 24,
  },
  icon: {
    marginEnd: 9,
  },
  propertyType: {
    color: theme.colors.primaryColor,
    marginBottom: 6,
  },
});
