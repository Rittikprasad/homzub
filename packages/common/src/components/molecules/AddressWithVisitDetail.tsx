import React from 'react';
import { StyleProp, StyleSheet, TouchableOpacity, View, ViewStyle } from 'react-native';
import { useTranslation } from 'react-i18next';
import { DateUtils } from '@homzhub/common/src/utils/DateUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';
import { PropertyAddress } from '@homzhub/common/src/components/molecules/PropertyAddress';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { VisitSlot } from '@homzhub/common/src/mocks/BookVisit';

interface IProps {
  primaryAddress: string;
  isPropertyOwner?: boolean;
  subAddress: string;
  startDate: string;
  endDate: string;
  comments?: string;
  isMissedVisit?: boolean;
  isCompletedVisit?: boolean;
  isRescheduleAll?: boolean;
  isFromProperty?: boolean;
  onPressSchedule?: () => void;
  navigateToAssetDetails: () => void;
  containerStyle?: StyleProp<ViewStyle>;
}

export const AddressWithVisitDetail = (props: IProps): React.ReactElement => {
  const {
    subAddress,
    primaryAddress,
    startDate,
    comments,
    isMissedVisit,
    isCompletedVisit,
    containerStyle = {},
    onPressSchedule,
    isRescheduleAll = false,
    isFromProperty = false,
    isPropertyOwner = false,
    navigateToAssetDetails,
  } = props;
  const { t } = useTranslation(LocaleConstants.namespacesKey.property);
  const dateTime = DateUtils.convertTimeFormat(startDate, 'YYYY-MM-DD HH');
  const time = VisitSlot.find((item) => item.from === Number(dateTime[1]));
  const textStyle = [styles.textColor, isMissedVisit && styles.missedColor];
  const isReschedule = isCompletedVisit || isMissedVisit || (!isCompletedVisit && !isMissedVisit && !isRescheduleAll);
  const isRescheduleAllVisible = isRescheduleAll && isPropertyOwner;

  const rescheduleView = (text: string): React.ReactElement => {
    return (
      <TouchableOpacity
        style={[styles.content, theme.viewport.width <= 375 && styles.extraMargin]}
        onPress={onPressSchedule}
      >
        <Icon name={icons.schedule} color={theme.colors.blue} size={20} />
        <Text type="small" style={styles.scheduleText}>
          {text}
        </Text>
      </TouchableOpacity>
    );
  };

  return (
    <View style={containerStyle}>
      {isFromProperty ? (
        <>
          {!!comments && (
            <Text type="small" textType="semiBold" style={styles.addressContainer}>
              {comments}
            </Text>
          )}
        </>
      ) : (
        <TouchableOpacity onPress={(): void => navigateToAssetDetails()}>
          <PropertyAddress
            primaryTextType="small"
            subAddress={subAddress}
            primaryAddress={primaryAddress}
            subAddressStyle={styles.subAddress}
            containerStyle={styles.addressContainer}
          />
        </TouchableOpacity>
      )}
      <Label type="regular" style={[styles.textColor, !comments && styles.detail]}>
        {t('visitDetails')}
      </Label>
      <View style={theme.viewport.width > 375 && styles.detailContainer}>
        <View style={styles.content}>
          <Text type="small" textType="semiBold" style={textStyle}>
            {DateUtils.getDisplayDate(dateTime[0], 'DD MMM')}
          </Text>
          <Icon
            name={icons.roundFilled}
            color={isMissedVisit ? theme.colors.error : theme.colors.darkTint3}
            size={8}
            style={styles.iconStyle}
          />
          <Text type="small" textType="semiBold" style={textStyle}>
            {time?.formatted}
          </Text>
        </View>
        {isReschedule && rescheduleView(isCompletedVisit ? t('newVisit') : t('reschedule'))}
        {isRescheduleAllVisible && rescheduleView(t('rescheduleAll'))}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  subAddress: {
    marginLeft: 0,
  },
  addressContainer: {
    marginVertical: 12,
  },
  textColor: {
    color: theme.colors.darkTint3,
  },
  detail: {
    marginTop: 12,
  },
  missedColor: {
    color: theme.colors.error,
  },
  detailContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  content: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  iconStyle: {
    marginTop: 6,
    marginHorizontal: 6,
  },
  scheduleText: {
    color: theme.colors.blue,
    marginStart: 6,
  },
  extraMargin: {
    marginTop: 6,
  },
});
