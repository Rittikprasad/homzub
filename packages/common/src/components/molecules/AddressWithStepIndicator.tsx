import React from 'react';
import { FlatList, StyleProp, StyleSheet, View, ViewStyle } from 'react-native';
import { useTranslation } from 'react-i18next';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { useDown, useViewPort } from '@homzhub/common/src/utils/MediaQueryUtils';
import { IBadgeInfo } from '@homzhub/mobile/src/navigation/interfaces';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Badge } from '@homzhub/common/src/components/atoms/Badge';
import { Image } from '@homzhub/common/src/components/atoms/Image';
import { ImagePlaceholder } from '@homzhub/common/src/components/atoms/ImagePlaceholder';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import { ITypographyProps } from '@homzhub/common/src/components/atoms/Typography';
import { PropertyAddressCountry } from '@homzhub/common/src/components/molecules/PropertyAddressCountry';
import { IAmenitiesData, PropertyAmenities } from '@homzhub/common/src/components/molecules/PropertyAmenities';
import { TypeOfPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { Attachment } from '@homzhub/common/src/domain/models/Attachment';
import { ILabelColor } from '@homzhub/common/src/domain/models/LabelColor';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { deviceBreakpoint } from '@homzhub/common/src/constants/DeviceBreakpoints';

interface IProps {
  steps: string[];
  currentIndex: number;
  isStepDone: boolean[];
  primaryAddress: string;
  subAddress: string;
  countryFlag: React.ReactElement;
  propertyType: string;
  icon?: string;
  onEditPress?: () => void;
  selectedPan?: string;
  badgeInfo?: IBadgeInfo;
  amenities?: IAmenitiesData[];
  onPressSteps: (index: number) => void;
  badgeStyle?: StyleProp<ViewStyle>;
  stepContainerStyle?: StyleProp<ViewStyle>;
  containerStyle?: StyleProp<ViewStyle>;
  stepIndicatorSeparatorStyle?: StyleProp<ViewStyle>;
  primaryAddressTextStyles?: ITypographyProps;
  subAddressTextStyles?: ITypographyProps;
  attachments?: Attachment[];
  topNode?: React.ReactElement;
  displayIndicator?: boolean;
}

export const AddressWithStepIndicator = (props: IProps): React.ReactElement => {
  const {
    steps,
    currentIndex,
    isStepDone,
    icon,
    selectedPan,
    badgeInfo,
    badgeStyle,
    amenities,
    stepContainerStyle = {},
    containerStyle = {},
    primaryAddress,
    subAddress,
    propertyType,
    onPressSteps,
    onEditPress,
    countryFlag,
    stepIndicatorSeparatorStyle,
    subAddressTextStyles,
    primaryAddressTextStyles,
    attachments,
    topNode,
    displayIndicator,
  } = props;

  const { width } = useViewPort();
  const { t } = useTranslation(LocaleConstants.namespacesKey.property);
  const isMobile = useDown(deviceBreakpoint.MOBILE);
  const isTablet = useDown(deviceBreakpoint.TABLET);
  const numberOfSteps = steps.length;
  const { rental, sell, informational, blue, green, darkTint7 } = theme.colors;
  const displayStepIndicator = displayIndicator ?? true;

  const badgeData = (): ILabelColor => {
    switch (selectedPan) {
      case TypeOfPlan.SELL:
        return {
          label: t('forSale'),
          color: sell,
        };
      case TypeOfPlan.MANAGE:
        return {
          label: t('inviteTenant'),
          color: informational,
        };
      case TypeOfPlan.RENT:
      default:
        return {
          label: t('forRental'),
          color: rental,
        };
    }
  };

  const renderIndicator = ({ item, index }: { item: string; index: number }): React.ReactElement => {
    const isCurrentStep = currentIndex === index;
    const stepDone = isStepDone[index];
    const iconColor = isCurrentStep ? blue : stepDone ? green : darkTint7;
    const labelStyle = [styles.stepLabel, isCurrentStep && styles.currentStepLabel, stepDone && styles.doneStepLabel];
    return (
      <View style={styles.indicatorView}>
        <Icon name={icons.roundFilled} size={30} color={iconColor} onPress={(): void => onPressSteps(index)} />
        <Label type="small" textType="semiBold" style={labelStyle}>
          {item}
        </Label>
      </View>
    );
  };

  const renderSeparator = (separatorData: { highlighted: boolean; leadingItem: string }): React.ReactElement => {
    let isDone = false;
    const indicatorSteps: string[] = steps;
    indicatorSteps.forEach((item, index) => {
      if (separatorData.leadingItem === item) {
        isDone = isStepDone[index];
      }
    });

    const calculateSeparatorWidth = (): number => {
      if (PlatformUtils.isMobile()) return (width - 220) / numberOfSteps;
      if (isMobile) return (width - 170) / numberOfSteps;
      if (isTablet) return (width - 230) / numberOfSteps;
      return (width - 600) / numberOfSteps;
    };

    const separatorWidth = { width: calculateSeparatorWidth() };

    return (
      <View style={[styles.separator, isDone && styles.doneSeparator, separatorWidth, stepIndicatorSeparatorStyle]} />
    );
  };

  const badge = badgeInfo ? { label: badgeInfo.title, color: badgeInfo.color } : badgeData();

  return (
    <View style={styles.container}>
      <View style={[styles.propertyDetailsContainer, containerStyle]}>
        {topNode && topNode}
        <View style={styles.propertyDetails}>
          {!isMobile && attachments && renderPropertyImage(attachments)}
          <View style={styles.propertyDetailsWrapper}>
            <View style={styles.topView}>
              <Label type="large" textType="regular" style={styles.propertyTypeStyle}>
                {propertyType}
              </Label>
              {icon && <Icon name={icon} size={23} color={blue} onPress={onEditPress} />}
              {(selectedPan || badgeInfo) && (
                <Badge title={badge.label.toUpperCase()} badgeColor={badge.color ?? ''} badgeStyle={badgeStyle} />
              )}
            </View>
            <PropertyAddressCountry
              primaryAddress={primaryAddress}
              subAddress={subAddress}
              countryFlag={countryFlag}
              containerStyle={styles.addressView}
              primaryAddressTextStyles={primaryAddressTextStyles}
              subAddressTextStyles={subAddressTextStyles}
            />
            {amenities && (
              <PropertyAmenities
                labelStyles={subAddressTextStyles}
                containerStyle={styles.amenities}
                contentContainerStyle={styles.amenitiesContentStyle}
                data={amenities}
                direction="row"
              />
            )}
          </View>
        </View>
      </View>
      {displayStepIndicator && (
        <View style={styles.stepWrapper}>
          <FlatList
            keyExtractor={(step): string => step}
            data={steps}
            renderItem={renderIndicator}
            horizontal
            showsHorizontalScrollIndicator={false}
            ItemSeparatorComponent={renderSeparator}
            style={[styles.listStyle, stepContainerStyle]}
          />
        </View>
      )}
    </View>
  );
};

const renderPropertyImage = (items: Attachment[]): React.ReactElement => {
  const data: Attachment | undefined = items.find(($0) => $0.mediaType === 'IMAGE');
  return (
    <>
      {data ? (
        <Image
          source={{
            uri: data.link,
          }}
          style={styles.propertyImage}
          resizeMode="contain"
        />
      ) : (
        <ImagePlaceholder containerStyle={styles.imagePlaceHolder} />
      )}
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    borderRadius: 4,
    backgroundColor: theme.colors.white,
  },
  propertyDetailsContainer: {
    padding: 16,
  },
  propertyDetailsWrapper: {
    flex: 1,
  },
  propertyDetails: {
    width: '100%',
    flexDirection: 'row',
  },
  propertyImage: {
    borderRadius: 4,
    overflow: 'hidden',
    marginRight: 16,
    height: 150,
    width: 300,
    resizeMode: 'cover',
    backgroundColor: theme.colors.darkTint1,
  },
  imagePlaceHolder: {
    height: 160,
    backgroundColor: theme.colors.disabled,
    marginRight: 16,
    paddingHorizontal: 10,
  },
  addressView: {
    marginVertical: 6,
  },
  indicatorView: {
    alignItems: 'center',
  },
  topView: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  separator: {
    backgroundColor: theme.colors.darkTint7,
    height: 2,
    marginTop: 10,
    opacity: 0.3,
  },
  doneSeparator: {
    backgroundColor: theme.colors.green,
  },
  stepWrapper: {
    width: '100%',
    borderTopWidth: 1,
    borderTopColor: theme.colors.background,
  },
  listStyle: {
    paddingVertical: 20,
    alignSelf: 'center',
  },
  stepLabel: {
    color: theme.colors.darkTint7,
  },
  currentStepLabel: {
    color: theme.colors.blue,
  },
  doneStepLabel: {
    color: theme.colors.green,
  },
  amenities: {
    marginTop: 10,
    marginBottom: 14,
    justifyContent: 'flex-start',
  },
  amenitiesContentStyle: {
    marginRight: 16,
  },
  propertyTypeStyle: {
    color: theme.colors.primaryColor,
  },
});
