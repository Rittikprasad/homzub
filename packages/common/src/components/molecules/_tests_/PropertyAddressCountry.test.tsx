import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PropertyAddressCountry } from '@homzhub/common/src/components/molecules/PropertyAddressCountry';

describe('PropertyAddressCountry', () => {
  const props = {
    primaryAddress: '2 BHK',
    subAddress: 'Delhi',
    countryFlag: '',
  };
  it('should match snapshot', () => {
    const wrapper = shallow(<PropertyAddressCountry {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
