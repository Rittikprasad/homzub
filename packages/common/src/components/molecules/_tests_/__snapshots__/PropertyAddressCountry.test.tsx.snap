// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyAddressCountry should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "column",
        "justifyContent": "space-between",
      },
      Object {},
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Typography
      fontWeight="semiBold"
      numberOfLines={1}
      size="regular"
      style={
        Array [
          Object {
            "alignItems": "center",
            "color": "#000",
            "flex": 1,
            "justifyContent": "center",
            "marginHorizontal": 12,
          },
          undefined,
        ]
      }
      variant="text"
    >
      2 BHK
    </Typography>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Typography
      fontWeight="regular"
      numberOfLines={2}
      size="small"
      style={
        Array [
          Object {
            "color": "#33495E",
            "flex": 1,
            "marginVertical": 6,
          },
          undefined,
        ]
      }
      variant="text"
    >
      Delhi
    </Typography>
  </View>
</View>
`;
